design/one-pager-generic-datasource.md

```diff
@@ -0,0 +1,186 @@
+* Status: Draft
+
+## Background
+
+Crossplane currently supports patching from composite resources to composed
+resources and vice versa. It is also possible to copy values between two
+composed resources inside one compositiion by patching the value from one
+composed back to the composite and then to the second composed resource.
+
+However, Crossplane currently does not provide a way to patch values from
+dynamic (aka environment-dependent) sources. 
+Compositions are pratically static right now, meaning, they behave the same
+regardless in which environment they are deployed.
+
+This becomes a big issue once you want to deploy the same composition in
+multiple environments. For example deploying AWS resources in two different VPCs
+requires different subnet IDs, VPC Ids, OIDC issuers etc.
+
+## Current workarounds
+
+### Build one composition per environment
+
+The simplest solution would be to build one composition per environment.
+While this indeed works, it becomes increasingly complex with the number of
+environments.
+
+### Use generic provider referencers
+
+While compositions itself are not dynamic, some providers serve some MRs that
+contain generic referencers. 
+
+* `provider-helm`: using [`spec.forProvider.valuesFrom.secretKeyRef`](https://doc.crds.dev/github.com/crossplane-contrib/provider-helm/helm.crossplane.io/Release/v1beta1@v0.10.0#spec-forProvider-valuesFrom-secretKeyRef)
+* `provider-kubernetes`: using [`spec.forProvider.references[].patchesFrom`](https://doc.crds.dev/github.com/crossplane-contrib/provider-kubernetes/kubernetes.crossplane.io/Object/v1alpha1@v0.3.0#spec-references-patchesFrom)
+
+However, none of these methods are Crossplane native and are not available to
+every provider. Furthermore, they would require an additional MR (here `Object`
+or `Release`) to deploy the composed resource increasing the complexity of the
+composition.
+
+## Goals
+
+The goal of this document is to define a proposal for extracting environment
+specific data considering the following requirements:
+
+1. Allow the creation of generic compositions that are able to load environment 
+specific parameters during compose time.
+2. Allow users to define these values on cluster level.
+3. Prevent users from escalating their rights on the cluster and read out fields
+from objects they shouldn't have access to.
+4. (Bonus) Allow package writers to include default values in configuration
+packages.
+
+## Proposed Solution
+
+Introduce a new Crossplane CRD `ConfigSet` that can store arbitrary data similar
+to K8s `ConfigMap` but supports more complex data types:
+
+```yaml
+---
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: ConfigSet
+metadata:
+  name: sample-config-set
+data:
+  key: value
+  complex:
+    a: b
+    c: d
+  list:
+    - a
+    - b
+    - c
+```
+
+Extend the composition CRD to allow patching from `ConfigSet`:
+
+```yaml
+        - type: FromConfigSetFieldPath
+          fromConfigSetRef:
```

Comment:
I'm a little bit on the fence about how to represent the reference. If we draw some similarities to `Combine` functionality, an alternative could be like the following:
```yaml
- type: FromConfigSet
  configSet:
    ref:
      name: sample-config-set
    fromFieldPath: data.key
  toFieldPath: spec.forProvider.manifest.data.static
  policy:
    fromFieldPath: Required
```
or keep using the same `fromFieldPath` at the top level.
```yaml
- type: FromConfigSet
  configSet:
    ref:
      name: sample-config-set
  fromFieldPath: data.key
  toFieldPath: spec.forProvider.manifest.data.static
  policy:
    fromFieldPath: Required
```
closer to the existing one
```yaml
- type: FromConfigSet
  configSetRef:
    name: sample-config-set
  fromFieldPath: data.key
  toFieldPath: spec.forProvider.manifest.data.static
  policy:
    fromFieldPath: Required
```

I'm essentially looking for ways to reduce the repetition of the words while keeping the structs contained to themselves and flexible enough for future iterations. @negz what are your thoughts?

Reply:
Good thought. I also felt that the design is pretty repetitive.

I would propose referencing the `ConfigSets` at the top of the Composition's spec:

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: cnp-virtualcontrolplane
spec:
  environment:
    configSetRefs:
      - name: sample-config-set
    # Alternatively, get it via labels
    configSetSelectors:
      - matchLabels:
          stage: prod
```

All found `ConfigSet`s are merged into one using strategic merging. Similar to Helm values.

Then you would just have to change the `type` in your patches:

```yaml
        - type: FromConfigSetFieldPath
          fromFieldPath: data.key
          toFieldPath: spec.forProvider.manifest.data.static
          policy:
            fromFieldPath: Required # Dont render if referenced field does not exist
```

Reply:
Once we want to use more than a singe configset ref, we'd need to make `environment` able to hold more than one, then I guess we'd have to ask which one to use under the patch instance, right? I feel like having `the identification of which ConfigSet to use` contained in the patch instance could allow us to be more flexible and less intrusive of other fields/features of `Composition`.

I feel like the following is the most similar to existing API conventions we have:
```yaml
- type: FromConfigSet # Or FromConfigSetToComposed? FromConfigSetToComposite in the future if users ask for it. Maybe FromConfigSet
  configSetRef:
    name: sample-config-set
  fromFieldPath: data.key
  toFieldPath: spec.forProvider.manifest.data.static
  policy:
    fromFieldPath: Required
```

Reply:
The similarity between `PatchSet` word used in `Composition` and the `ConfigSet` is something I just realized, though not sure if it's necessarily a bad thing.

Reply:
I updated the [draft implementation](https://github.com/crossplane/crossplane/pull/3007/files#diff-5d7910f2ba87282df3f14400d5ec4d39c262e3b7b444b7889c9887c2dbf92162R104) that includes merging all referenced `ConfigSets` into one and make the patch agnostic of the source. This feels the most natural to me and I would propose to go forward with this.

Reply:
@MisterMX I'm onboard with merging referenced ConfigSets into one set of data (e.g. when the `environmentConfigSelector` matches multiple `EnvironmentConfigs`). I think that's a related but separate concern from whether we'd want to let folks build multiple named 'config sets' within a Composition.

Allowing multiple named config sets is probably unnecessary complexity when each set can be merged from multiple inputs, but I'm tempted to push for it anyhow mostly because it makes the API work very similarly to the existing "patch sets" functionality.

Reply:
>  (e.g. when the environmentConfigSelector matches multiple EnvironmentConfigs)

Do we want to select multiple ConfigSets through one label selector? I think this might be untransparent to the user because the order of items returned by `kube.List()` can be quite different. I would suggest to only select the first and require the user to make another selection if he needs more.

Reply:
> Do we want to select multiple ConfigSets through one label selector?

Oh, I got mixed up and thought this was what you had been proposing. I agree that it could be confusing.

So would something like this be what you were thinking:

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: example
spec:
  patchSets:
  - name: metadata
    patches:
    - type: FromCompositeFieldPath
      fromFieldPath: metadata.labels[some-important-label]
  configSets:
    # The 'name' field is now just informational, to distinguish entries.
    # We could omit it if we wanted to.
  - name: staging
    environmentConfigRef:
    - name: staging
  - name: west-coast
    environmentConfigRef:
    - name: west-coast
  resources:
  - name: example
    base:
      apiversion: example.org/v1
      kind: Foo
      spec:
        forProvider:
          bars: 42
    patches:
    - type: PatchSet
      patchSetName: metadata
    # FromConfigSetFieldPath _only_ patches from ConfigSet -> Composed resource.
    # It doesn't make sense for patching from ConfigSet -> Composite resource to
    # be associated with a particular composed resource template.
    - type: FromConfigSetFieldPath
      fromFieldPath: data.bazes
      toFieldPath: spec.forProvider.bazes
      policy:
        fromFieldPath: Required
  # Introduce a new top-level 'composite' field. This field can be used to
  # mutate the composite resource in ways that don't have anything to do with
  # a particular composed resource. In theory we could avoid this by patching
  # from configSet -> composed -> composite but that seems awkward.
  composite:
    patches:
    - type: FromConfigSetFieldPath
      fromFieldPath: data.bangs
      toFieldPath: status.bangs
    # We could move this recently added alpha field here too
    publishConnectionDetailsWithStoreConfigRef:
      name: default
```

Note that there's no longer a reference to a particular `configSetName`, it just references the merged data.

Reply:
I'm going to resolve this thread and respond to https://github.com/crossplane/crossplane/pull/3008#issuecomment-1135756546, which touches on the same topic.